### DISCUSSION TITLE
Docs say it's possible, but no one says how - Polymorphic identities using foreign keys other than the primary key

### DISCUSSION THREAD

**Discussion by talglobus** [sqlalchemy/sqlalchemy#11149] [Usage Questions]
I've spent a few hours tracking down git blames on one obscure line in the [2.0 docs](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#joined-table-inheritance):

> In a polymorphic setup, it is most common that the foreign key constraint is established on the same column or columns as the primary key itself, however this is not required; a column distinct from the primary key may also be made to refer to the parent via foreign key. The way that a JOIN is constructed from the base table to subclasses is also directly customizable, however this is rarely necessary.

This line was added in [this commit](https://github.com/sqlalchemy/sqlalchemy/commit/295441bab3051e96e7651047acb02cabca9ac0f5) by @Quelklef from 2020, though @zzzeek [added](https://github.com/sqlalchemy/sqlalchemy/commit/6194483b73c94565aee62c5867a251a1841db5a9) this line in 2017:

> It is most common that the foreign key constraint is established on the same column or columns as the primary key itself, however this is not required; a column distinct from the primary key may also be made to refer to the parent via foreign key.

I traced this description through several iterations all the way back to the creation of the file in [this commit](https://github.com/sqlalchemy/sqlalchemy/commit/4e24d6cb01c58d9617d5e22e1068b2d164e4a3db#diff-c62c136c91ef8dd2c66c4ce3e20385f61f964e1b698f44a7b65055fa41074273) from @zzzeek all the way back in 2010, with the following:

> Each table also must contain a primary key column (or columns), and in most cases a foreign key reference to the parent table. It is standard practice that the same column is used for both of these roles, and that the column is also named the same as that of the parent table. However this is optional in SQLAlchemy; separate columns may be used for primary key and parent-relationship, the column may be named differently than that of the parent, and even a custom join condition can be specified between parent and child tables instead of using a foreign key:

```
engineers = Table('engineers', metadata,
   Column('employee_id', Integer, ForeignKey('employees.employee_id'), primary_key=True),
   Column('engineer_info', String(50)),
)

managers = Table('managers', metadata,
   Column('employee_id', Integer, ForeignKey('employees.employee_id'), primary_key=True),
   Column('manager_data', String(50)),
)
```

**_But as far as I can tell, nowhere in those 14 years of history did anyone explain how to actually do so._**

When I try to do the following, I get this error from Alembic:
```
/Users/me/foo/migrations/models.py:132: SAWarning: Implicitly combining column user.id with column business_user.id under attribute 'id'.  Please configure one or more attributes for these same-named columns explicitly.
  class BusinessUser(User):
/Users/me/foo/migrations/models.py:168: SAWarning: Implicitly combining column user.id with column consumer_user.id under attribute 'id'.  Please configure one or more attributes for these same-named columns explicitly.
  class ConsumerUser(User):
```

Here are the models
```
class User(Base):
    __tablename__ = "user"

    id: Mapped[str] = MappedColumn(
        primary_key=True
    )
    user_type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "user",
        "polymorphic_on": "user_type"
    }


class BusinessUser(User):
    __tablename__ = "business_user"

    id: Mapped[int] = MappedColumn(
        primary_key=True, default=generate_id_with_prefix("busr")
    )
    user_id: Mapped[str] = MappedColumn(
        ForeignKey("user.id"), nullable=False
    )

    __mapper_args__ = {
        'polymorphic_identity': 'business',
    }


class ConsumerUser(User):
    __tablename__ = "consumer_user"

    id: Mapped[int] = MappedColumn(
        primary_key=True, default=generate_id_with_prefix("cusr")
    )
    user_id: Mapped[str] = MappedColumn(
        ForeignKey("user.id"), nullable=False
    )

    __mapper_args__ = {
        'polymorphic_identity': 'consumer',
    }
```

I have naming conventions for my IDs, and losing the naming scheme separating consumer users from business users would make support much more difficult. If anyone could let me know how to use a different column to reference the parent from the ID, I'd really appreciate it.

**Comment by zzzeek** [comment]
your foreign key setup is fine.  the warning you are seeing refers to your "id" attribute, and the issue is that if you have a different "id" in `business_user.id` versus `user.id`, what value is returned if you request `BusinessUser.id` ?   Therefore you need to disambiguate these, such as like this:

```py
class BusinessUser(User):
    __tablename__ = "business_user"

    business_id: Mapped[int] = MappedColumn(
        "id",
        primary_key=True
    )
    user_id: Mapped[str] = MappedColumn(
        ForeignKey("user.id"), nullable=False
    )

    __mapper_args__ = {
        'polymorphic_identity': 'business',
    }

```

there is no need to change table-level naming conventions, just class level naming conventions

**Reply by talglobus** [reply]
Is there a way of overriding the `User` primary key that SQLAlchemy will understand? We actually never need the `User.id`, only the other fields on the `User`, and if it were up to us the `User` would inherit the `id` from the child class, using the naming convention of the child

**Reply by zzzeek** [reply]
OK then do it the other way around:

```
class User(Base):
    __tablename__ = "user"

    _user_id: Mapped[str] = MappedColumn(
        "id",
        primary_key=True
    )
    user_type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "user",
        "polymorphic_on": "user_type"
    }

```
